Tone_Alarm: Added ability to use timer 1 and timer 8 for the tone alarm driver.

ksschwabe opened the pull request:
I added the ability to use the advanced timers TIM1 and TIM8 in the tone_alarm driver. Since these are advanced timers, they have an additional register (break and dead-time register). The main output enable (MOE) bit in this register needs to be set to have the timer output tied to the pin.

This PR doesn't affect any of the current working of the tone alarm driver for any of the current boards, it just allows one to make use of timers 1 and 8, if someone wants to use the tone_alarm driver on another board that has a buzzer on timer 1 or 8.

davids5 commented:
@ksschwabe 

Thank you for the changes.

There is a very subtle hint in the file that the tone alarm should be on a general (generic) timer. 
```
# error Must set TONE_ALARM_TIMER to a generic timer in order to use this driver.
```
I think this is due to the fact that the majority of the boards use timer 1 or 8 for HRT.

To avoid the conflict it may be prudent to add something like
```
#if defined(TONE_ALARM_TIMER)  && defined(HRT_TIMER)
# if  TONE_ALARM_TIMER == HRT_TIMER
#   error TONE_ALARM_TIMER and HRT_TIMER must use different timers. 
# endif
#endif
```
and also change the ```# error Must set TONE_ALARM_TIMER to a generic timer in order to use this driver``` message

Also were you able to test this on HW and check the frequency? 

I ask because I have not reviewed the math and there are 2 factors to consider: 

1) Any APB2 clocked timers (1,8,9-11)  are at 2 X the clock rate of the APB1 clocked timers (with the board clocks settings and state of TIMPRE).  Because the code had support for timers 9-11 your change would not introduce this as a new error but it would be good to know that (1,8,9-11) work.

2) To further complicate things, per AN4013  Timer 2,5 on the F4 are 32 bits.
![image](https://cloud.githubusercontent.com/assets/1945821/9658929/bdf9aad6-51e8-11e5-9e8b-e8333fde4751.png)

So if the the calculation overflows on timer 2 or 5 into the upper 16 bits of the 32 bit registers things still work because the ```#define REG(_reg)	(*(volatile uint32_t *)(TONE_ALARM_BASE + _reg))``` treats all registers as 32 bits. (that is another story)

The math may be fine - but it is worth checking it. 

Would you be able to to scope it or hook up a speaker?

ksschwabe commented:
@davids5 

Yes, normally the HRT timer is on TIM1 or TIM8. I have added the compile time check as requested and changed the error message. Great idea :+1: 

I was porting a board with an STM32F100 microcontroller and was using the PX4IO as a starting point, but it had the buzzer on TIM1, hence why I made this change to use the tone_alarm on one of the advanced timers.

On the FMUV1 board, the buzzer pin PC8 can be driven by TIM3 and TIM8, so it was easy to test whether TIM8 would also work there. I scoped the the output using TIM3 and TIM8 and it gave me the same frequency and pulse width. The different clock speeds on APB1 and APB2 are taken into account in the function note_to_divisor(unsigned note), where it uses `TONE_ALARM_CLOCK` to calculate the divisor.
```
unsigned
ToneAlarm::note_to_divisor(unsigned note)
{
	// compute the frequency first (Hz)
	float freq = 880.0f * expf(logf(2.0f) * ((int)note - 46) / 12.0f);

	float period = 0.5f / freq;

	// and the divisor, rounded to the nearest integer
	unsigned divisor = (period * TONE_ALARM_CLOCK) + 0.5f;

	return divisor;
}
```

Regarding the 32 bit timers, I don't think we have an overflow problem. At least I hope we don't. I have used TIM2 for the buzzer on one of my boards and I have never had any problems with it, so I think it is fine, but I haven't yet gone through all of the maths.

davids5 commented:
@ksschwabe 

Awesome work - thank you for for scoping it.
